CloneTest/Program.cs

```diff
@@ -1,5 +1,139 @@
+        for (int i = 0; i < startingHubs.Count(); i++) {
+            if (startingHubs[i].Item1 == endingHubs[i].Item1 && startingHubs[i].Item2 == endingHubs[i].Item2) {
+                startingHubs.RemoveAt(i);
+                endingHubs.RemoveAt(i);
+                i--;
+            }
+        }
+
+        //Liczenie hubów o danych koordynatach
+        var startHubsCount = CountHubs(startingHubs);
+        var endHubsCount = CountHubs(endingHubs);
+
+        //Sortowanie listy hubow pod względem liczby wystąpień + Wypisanie raportu
+        OutputReport(Quicksort.SortList(startHubsCount, 0, startHubsCount.Count() - 1), "poczatkowe");
+        OutputReport(Quicksort.SortList(endHubsCount, 0, endHubsCount.Count() - 1), "koncowe");
+
+    }
+
+
+    protected static List<Tuple<int, int, double>> ListRandomAssign(double[,] list, int xMax, int yMax, int nPeople) {
+        Random rnd = new();
+        var arr = new List<Tuple<int, int, double>>();
+
+        for (int i = 0; i < nPeople; i++) {  //Wybór losowych komórek z tablicy gęstości i zapisanie do listy hubsDensity
+            int xRnd = rnd.Next(0, xMax);
+            int yRnd = rnd.Next(0, yMax);
+            arr.Add(Tuple.Create(xRnd, yRnd, list[xRnd, yRnd]));
+        }
+        return arr;
+    }
+
+    protected static List<double> CdfCalculate(List<Tuple<int, int, double>> arr, int nPeople) { //Obliczenie dystrybuant i zapisanie ich do zwracanej listy (cdfOut)
+        double temp = 0;
+        var cdfOut = new List<double>();
+
+        for (int i = 0; i < nPeople; i++) {
+            temp += arr[i].Item3;
+            cdfOut.Add(temp);
+        }
+        return cdfOut;
+
+    }
+
+    protected static List<Tuple<int, int>> ChooseHubs(List<double> cdf, List<Tuple<int, int, double>> hubs, int nHubs) {
+        var list = new List<Tuple<int, int>>();
+        
+        for (int i = 0; i < nHubs; i++) {
+            double cdfTarget = RandomDouble(cdf[0], cdf.Last());  //Losowanie wartości gestosci (na podstawie dystrybuanty)
+            int cdfFound = BinarySearch.SearchClosest(cdf, cdfTarget); //Wyszukiwanie hub'a spełniającego warunek cdf[i] < cdfTarget < cdf[i+1]
+
+            list.Add(Tuple.Create(hubs[cdfFound].Item1, hubs[cdfFound].Item2)); // Przepisanie wartości x i y wylosowanego huba do listy zwracanej
+        }
+        return list;
+    }
+
+    protected static List<Tuple<int, int, int>> CountHubs(List<Tuple<int, int>> arr) {
+        var outList = new List<Tuple<int, int, int>>();
+
+        while(arr.Count() != 0) {
+            var temp = Tuple.Create(arr[0].Item1, arr[0].Item2);
+            int count = 0;
+            for (int j = 0; j < arr.Count(); j++) {
+                if (arr[j].Equals(temp)) {
+                    count++;
+                    arr.RemoveAt(j);
+                    j--;
+                }
+            }
+            outList.Add(Tuple.Create(temp.Item1, temp.Item2, count));
+        }
```

Comment:
Pętla w pętli źle współgra z wydajnością. Skorzystaj w tej sytuacji z `Dictionary<Tuple<int, int, int, int>, int>`, gdzie kluczem są koordynaty połączenia i wartością to ile razy się pojawiły. Kod powinien w tej sytuacji wyglądać mniej więcej tak:
```csharp
var dict = new Dictionary<Tuple<int, int, int, int>, int>();
foreach(var i in arr){
    if(dict.ContainsKey(i)) dict[i]++;
    else dict.Add(i, 1);
}
```